pkg/network/ebpf/c/protocols/tls/tls.h

```diff
@@ -11,122 +10,420 @@
+static __always_inline bool read_tls_record_header(struct __sk_buff *skb, __u64 nh_off, tls_record_header_t *tls_hdr) {
+    __u32 skb_len = skb->len;
+
+    // Ensure there's enough space for TLS record header
+    if (nh_off + sizeof(tls_record_header_t) > skb_len)
         return false;
-    }
-    // Checking the tls record header length is greater than the tls hello message header length.
-    if (hdr->length < sizeof(tls_hello_message_t)) {
+
+    // Read TLS record header
+    if (bpf_skb_load_bytes(skb, nh_off, tls_hdr, sizeof(tls_record_header_t)) < 0)
         return false;
-    }
 
-    // Getting the tls hello message header.
-    tls_hello_message_t msg = *(tls_hello_message_t *)(buf + sizeof(tls_record_header_t));
-    // If the message is not a CLIENT_HELLO or SERVER_HELLO, we don't attempt to classify.
-    if (msg.handshake_type != TLS_HANDSHAKE_CLIENT_HELLO && msg.handshake_type != TLS_HANDSHAKE_SERVER_HELLO) {
+    // Convert fields to host byte order
+    tls_hdr->version = bpf_ntohs(tls_hdr->version);
+    tls_hdr->length = bpf_ntohs(tls_hdr->length);
+
+    // Validate version and length
+    if (!is_valid_tls_version(tls_hdr->version))
         return false;
-    }
-    // Converting the fields to host byte order.
-    __u32 length = msg.length[0] << 16 | msg.length[1] << 8 | msg.length[2];
-    // TLS handshake message length should be equal to the record header length minus the size of the hello message
-    // header.
-    if (length + TLS_HELLO_MESSAGE_HEADER_SIZE != hdr->length) {
+    if (tls_hdr->length > TLS_MAX_PAYLOAD_LENGTH)
         return false;
-    }
 
-    msg.version = bpf_ntohs(msg.version);
-    return is_valid_tls_version(msg.version) && msg.version >= hdr->version;
+    // Ensure we don't read beyond the packet
+    if (nh_off + sizeof(tls_record_header_t) + tls_hdr->length > skb_len)
+        return false;
+
+    return true;
 }
 
-// is_tls checks if the given buffer is a valid TLS record header. We are
-// currently checking for two types of record headers:
-// - TLS Handshake record headers
-// - TLS Application Data record headers
-static __always_inline bool is_tls(const char *buf, __u32 buf_size, __u32 skb_len) {
-    if (buf_size < sizeof(tls_record_header_t)) {
+// is_tls checks if the packet is a TLS packet and reads the TLS record header
+static __always_inline bool is_tls(struct __sk_buff *skb, __u64 nh_off, tls_record_header_t *tls_hdr) {
+    // Use the helper function to read and validate the TLS record header
+    if (!read_tls_record_header(skb, nh_off, tls_hdr))
+        return false;
+
+    // Validate content type
+    if (tls_hdr->content_type != TLS_HANDSHAKE && tls_hdr->content_type != TLS_APPLICATION_DATA)
         return false;
+
+    return true;
+}
+
+static __always_inline int parse_tls_handshake_header(struct __sk_buff *skb, __u64 *offset, __u32 skb_len, __u32 *handshake_length, __u16 *protocol_version) {
```

Comment:
`offset` is being incremented by these helpers so that some code can be shared by the `client` and `server` hello packet inspection. `skb_len` can be fetched from `skb`, I'll remove the passing around of that value

Reply:
this is no longer applicable as we're using `data_end` now